src/webgpu/shader/validation/types/textures.spec.ts

```diff
@@ -0,0 +1,100 @@
+export const description = `
+Validation tests for various texture types in shaders.
+
+TODO:
+- Sampled Texture Types
+- Multisampled Texture Types
+- External Sampled Texture Types
+- Depth Texture Types
+- Sampler Types
+`;
+
+import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import {
+  isTextureFormatUsableAsStorageFormat,
+  kAllTextureFormats,
+  kColorTextureFormats,
+  kTextureFormatInfo,
+} from '../../../format_info.js';
+import { getPlainTypeInfo } from '../../../util/shader.js';
+import { ShaderValidationTest } from '../shader_validation_test.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+g.test('texel_formats')
+  .desc(
+    'Test channels and channel format of various texel formats when using as the storage texture format'
+  )
+  .params(u =>
+    u
+      .combine('format', kColorTextureFormats)
+      .filter(p => kTextureFormatInfo[p.format].color.storage)
+      .beginSubcases()
+      .combine('shaderScalarType', ['f32', 'u32', 'i32', 'bool', 'f16'] as const)
+  )
+  .beforeAllSubcases(t => {
+    if (t.params.shaderScalarType === 'f16') {
+      t.selectDeviceOrSkipTestCase({ requiredFeatures: ['shader-f16'] });
+    }
+
+    if (!isTextureFormatUsableAsStorageFormat(t.params.format, t.isCompatibility)) {
+      t.skip('storage usage is unsupported');
+    }
+  })
+  .fn(t => {
+    const { format, shaderScalarType } = t.params;
+    const info = kTextureFormatInfo[format];
+    const validShaderScalarType = getPlainTypeInfo(info.color.type);
+    const shaderValueType = `vec4<${shaderScalarType}>`;
+    const wgsl = `
+    @group(0) @binding(0) var tex: texture_storage_2d<${format}, read>;
+    @compute @workgroup_size(1) fn main() {
+        let v : ${shaderValueType} = textureLoad(tex, vec2u(0));
+        _ = v;
+    }
+`;
+    t.expectCompileResult(validShaderScalarType === shaderScalarType, wgsl);
+  });
+
+g.test('texel_formats,as_value')
+  .desc('Test that texel format cannot be used as value')
+  .fn(t => {
+    const wgsl = `
+    @compute @workgroup_size(1) fn main() {
+        var i = rgba8unorm;
+    }
+`;
+    t.expectCompileResult(false, wgsl);
+  });
+
+g.test('storage_texture_types')
+  .desc(
+    `Test that for texture_storage_xx<format, access>
+- format must be an enumerant for one of the texel formats for storage textures
+- access must be an enumerant for one of the access modes
+`
+  )
+  .params(u =>
+    u
+      .combine('access', ['read', 'write', 'read_write'] as const)
+      .combine('format', kAllTextureFormats)
+      // bgra8unorm-storage feature is tested at webgpu,api,validation,texture,bgra8unorm_storage
```

Comment:
Overlap is fine. Suggest removing.

Reply:
Remove filter (add bgra8unorm).
Seems to fail for all access (when bgra8unorm-storage is not requried)

Reply:
> Seems to fail for all access (when bgra8unorm-storage is not requried)

Sorry, please can you elaborate a little? You believe you've found an implementation bug?

Reply:
Please can you use `run-cts --dump-shaders` and file a tint bug with the WGSL that you think is broken? I'll investigate.

Reply:
> So I think the problem is that the test uses the API's expectations for whether this is valid, instead of WGSL's.

Do you refer to `bgra8unorm_storage.spec.ts` or "the test added in this CL" for "the tests"?

Reply:
FYI the dumpped wgsl for `webgpu:shader,validation,types,textures:storage_texture_types:access="read";format="bgra8unorm"`
```
// Dumped WGSL:
@group(0) @binding(0) var tex: texture_storage_2d<bgra8unorm, read>;
```

Reply:
For this test, it uses info hardcoded in `format_info.ts` for storage and readWriteStorage usage.
for bgra8unorm, both storage and readWriteStorage are false. If WGSL should always accept the format as storage, that information from `format_info.ts` seems to include API restriction.

Given that l think for bgra8unorm we should skip it here, and let `bgra8unorm_storage.spec.ts` to test for this special case. WDYT?

Reply:
I meant the tests added here, but it seems that `bgra8unorm_storage.spec.ts` also expects shader modules containing `bgra8unorm_storage` storage textures to fail when `bgra8unorm-storage` is not present. That doesn't seem quite right to me, since I don't see anywhere in the API spec that imposes format validation on the shader module itself.

Reply:
Update the test with reinterpreted spec. Now the only case failing is bgra8unorm. PTAL